gadget/filesystemimage_test.go

```diff
@@ -0,0 +1,375 @@
+		mkfsCalled = true
+		return nil
+	}
+
+	restore := gadget.MockMkfsHandlers(map[string]gadget.MkfsFunc{
+		"happyfs": mkfsHappyFs,
+	})
+	defer restore()
+
+	fiw, err := gadget.NewFilesystemImageWriter(s.content, ps, s.work)
+	c.Assert(err, IsNil)
+
+	err = fiw.Write(img, cb)
+	c.Assert(err, IsNil)
+	c.Assert(cbCalled, Equals, true)
+	c.Assert(mkfsCalled, Equals, true)
+	// nothing left in temporary staging location
+	matches, err := filepath.Glob(s.work + "/*")
+	c.Assert(err, IsNil)
+	c.Assert(matches, HasLen, 0)
+}
+
+func (s *filesystemImageTestSuite) TestPostStageOptional(c *C) {
+	var mkfsCalled bool
+	mkfsHappyFs := func(imgFile, label, contentsRootDir string) error {
+		mkfsCalled = true
+		return nil
+	}
+
+	restore := gadget.MockMkfsHandlers(map[string]gadget.MkfsFunc{
+		"happyfs": mkfsHappyFs,
+	})
+	defer restore()
+
+	fiw, err := gadget.NewFilesystemImageWriter(s.content, s.psTrivial, s.work)
+	c.Assert(err, IsNil)
+
+	img := s.imgForPs(c, s.psTrivial)
+
+	err = fiw.Write(img, nil)
+	c.Assert(err, IsNil)
+	c.Assert(mkfsCalled, Equals, true)
+}
+
+func (s *filesystemImageTestSuite) TestChecksImage(c *C) {
+	cb := func(rootDir string, cbPs *gadget.PositionedStructure) error {
+		return errors.New("unexpected call")
+	}
+
+	mkfsHappyFs := func(imgFile, label, contentsRootDir string) error {
+		return errors.New("unexpected mkfs call")
+	}
+
+	restore := gadget.MockMkfsHandlers(map[string]gadget.MkfsFunc{
+		"happyfs": mkfsHappyFs,
+	})
+	defer restore()
+
+	fiw, err := gadget.NewFilesystemImageWriter(s.content, s.psTrivial, s.work)
+	c.Assert(err, IsNil)
+
+	img := filepath.Join(s.dir, "img")
+
+	// no image file
+	err = fiw.Write(img, cb)
+	c.Assert(err, ErrorMatches, "cannot stat image file: .*/img: no such file or directory")
+
+	// image file smaller than expected
+	makeSizedFile(c, img, s.psTrivial.Size/2, nil)
+
+	err = fiw.Write(img, cb)
+	c.Assert(err, ErrorMatches, fmt.Sprintf("size of image file %v is different from declared structure size %v", s.psTrivial.Size/2, s.psTrivial.Size))
+}
+
+func (s *filesystemImageTestSuite) TestPostStageError(c *C) {
+	cb := func(rootDir string, cbPs *gadget.PositionedStructure) error {
+		return errors.New("post stage exploded")
+	}
+
+	mkfsHappyFs := func(imgFile, label, contentsRootDir string) error {
+		return errors.New("unexpected mkfs call")
+	}
+	restore := gadget.MockMkfsHandlers(map[string]gadget.MkfsFunc{
+		"happyfs": mkfsHappyFs,
+	})
+	defer restore()
+
+	fiw, err := gadget.NewFilesystemImageWriter(s.content, s.psTrivial, s.work)
+	c.Assert(err, IsNil)
+
+	img := s.imgForPs(c, s.psTrivial)
+
+	err = fiw.Write(img, cb)
+	c.Assert(err, ErrorMatches, "post stage callback failed: post stage exploded")
+}
+
+func (s *filesystemImageTestSuite) TestMkfsError(c *C) {
+	mkfsHappyFs := func(imgFile, label, contentsRootDir string) error {
+		return errors.New("will not mkfs")
+	}
+	restore := gadget.MockMkfsHandlers(map[string]gadget.MkfsFunc{
+		"happyfs": mkfsHappyFs,
```

Comment:
heh, this is "unhappyfs" now :)  (just kidding, I see it can't be changed without changing psTrivial as well)